src/client/conn.rs

```diff
@@ -0,0 +1,456 @@
+    B::Item: AsRef<[u8]>,
+{
+    /// Sends a `Request` on the associated connection.
+    ///
+    /// Returns a future that if successful, yields the `Response`.
+    pub fn send_request(&mut self, req: Request<B>) -> ResponseFuture {
+        let inner = self.send_request_retryable(req).map_err(|e| {
+            let (err, _orig_req) = e;
+            err
+        });
+        ResponseFuture {
+            inner: Box::new(inner),
+        }
+    }
+
+    pub(super) fn close(&self) {
+        self.dispatch.cancel();
+    }
+
+    //TODO: replace with `impl Future` when stable
+    pub(crate) fn send_request_retryable(&mut self, req: Request<B>) -> Box<Future<Item=Response, Error=(::Error, Option<(::proto::RequestHead, Option<B>)>)>> {
+        let (head, body) = proto::request::split(req);
+        let inner = match self.dispatch.send((head, body)) {
+            Ok(rx) => {
+                Either::A(rx.then(move |res| {
+                    match res {
+                        Ok(Ok(res)) => Ok(res),
+                        Ok(Err(err)) => Err(err),
+                        // this is definite bug if it happens, but it shouldn't happen!
+                        Err(_) => panic!("dispatch dropped without returning error"),
+                    }
+                }))
+            },
+            Err(req) => {
+                debug!("connection was not ready");
+                let err = ::Error::new_canceled(None);
+                Either::B(future::err((err, Some(req))))
+            }
+        };
+        Box::new(inner)
+    }
+}
+
+/*
+impl<T, B> Service for SendRequest<T, B> {
+    type Request = Request<B>;
+    type Response = Response;
+    type Error = ::Error;
+    type Future = ResponseFuture;
+
+    fn call(&self, req: Self::Request) -> Self::Future {
+
+    }
+}
+*/
+
+impl<B> fmt::Debug for SendRequest<B> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("SendRequest")
+            .finish()
+    }
+}
+
+// ===== impl Connection
+
+impl<T, B> Connection<T, B>
+where
+    T: AsyncRead + AsyncWrite,
+    B: Stream<Error=::Error> + 'static,
+    B::Item: AsRef<[u8]>,
+{
+    /*
+    pub fn into_inner(self) -> T {
+
+    }
+    */
+
+    /// Return the inner IO object, and additional information.
+    pub fn into_parts(self) -> Parts<T> {
+        let (io, read_buf) = self.inner.into_inner();
+        Parts {
+            io: io,
+            read_buf: read_buf,
+            _inner: (),
+        }
+    }
+
+    /// Poll the connection for completion, but without calling `shutdown`
+    /// on the underlying IO.
+    ///
+    /// This is useful to allow running a connection while doing an HTTP
+    /// upgrade. Once the upgrade is completed, the connection would be "done",
+    /// but it is not desired to actally shutdown the IO object. Instead you
+    /// would take it back using `into_parts`.
+    pub fn poll_without_shutdown(&mut self) -> Poll<(), ::Error> {
```

Comment:
Couldn't poll know that the connection's upgraded and not shut it down?

Reply:
It could, but that's probably a trap, since you may not choose to do anything special with a `Connection`, and just call `exe.spawn(conn.map_err(|e| debug!("conn error: {}", e))`. If you did that, and `poll` did what `poll_with_shutdown` does in this PR, then the transport would never have `shutdown` called.

If `Connection::Item` were `Parts`, then that could probably be easier, since then you'd need to handle the `Parts` in some way (even if just dropping) when spawning the future. But, in order to do that, `Connection` needs to hold the internal state in an `Option`, so it can remove it, and thus every poll requires a `self.inner.as_mut().unwrap()`. I figured by providing `poll_without_shutdown`, `Connection::poll` isn't bothered, and anyone could wrap `poll_with_shutdown` and an `Option` if they wanted that kind of API.

Does that reasoning seem convincing?

Reply:
FWIW storing an Option and .as_mut().unwrap()'ing it is super common in futures/stream combinators so it wouldn't be that crazy. Seems fine to have two methods though.